### PR TITLE
add http2 testing interop server uses small data frames and padding

### PR DESCRIPTION
**Pull request by apolcyn**
Prompted by/meant to verify https://github.com/grpc/grpc-go/pull/1076

Adds a new server impl to python testing http2 server for handling interop `UnaryCalls` - it sends the responses across series of small data frames that use padding - meant to show things like not accounting for padding in flow control. 

This adds option to the interop script to run existing interop client `LargeUnary` methods against this test server. 

Currently test status: 
* java passes
* cc @ctiller core-based transports fail fast with `Error status code: 2 (expected: 0), message: Stream removed` 
* go passes with https://github.com/grpc/grpc-go/pull/1076, but go master times out, appears to run out of flow control quota.

### CODE REVIEW

File: test/http2_test/http2_base_server.py [hunk 1]
```diff
@@ -129,30 +129,34 @@ def send_reset_stream(self):
...
-      chunk_size = min(lfcw, _READ_CHUNK_SIZE)
+      chunk_size = min(lfcw - padding_bytes, read_chunk_size)
       bytes_to_send = min(chunk_size, self._send_remaining[stream_id])
-      logging.info('flow_control_window = %d. sending [%d:%d] stream_id %d' %
-                    (lfcw, self._send_offset, self._send_offset + bytes_to_send,
-                    stream_id))
+      logging.info('flow_control_window = %d. sending [%d:%d] stream_id %d. includes %d total padding bytes' %
+                    (lfcw, self._send_offset, self._send_offset + bytes_to_send + padding_bytes,
+                    stream_id, padding_bytes))
+      if bytes_to_send + padding_bytes > 16384:
```
Comment (makdharma) [inline test/http2_test/http2_base_server.py]: Please replace the magic number 16384 with a define for readability

File: test/http2_test/test_data_frame_padding.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    logging.info('on window update. Stream id: %s. Delta: %s' % (event.stream_id, event.delta))
+    self._total_updates[event.stream_id] += event.delta
+    total = self._total_updates[event.stream_id]
+    logging.info('... - total updates for stream %d : %d' % (event.stream_id, total))
+    p = []
+    if event.stream_id == 0:
+      p = self._cur_sending
+    else:
+      p = [event.stream_id]
+    for p in self._cur_sending:
```
Comment (makdharma) [inline test/http2_test/test_data_frame_padding.py]: You create a list named 'p' but reuse the name again. What is the purpose of assigning anything to p before the for loop?

File: test/http2_test/test_data_frame_padding.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    self._cur_sending.remove(stream_id)
+    self._base_server.on_send_done_default(stream_id)
+
+  def on_request_received(self, event):
+    self._base_server.on_request_received_default(event)
+    logging.info('on request received. Stream id: %s.' % event.stream_id)
+    self._total_updates[event.stream_id] = 0
+
+  def on_window_update(self, event):
+    logging.info('on window update. Stream id: %s. Delta: %s' % (event.stream_id, event.delta))
```
Comment (makdharma) [inline test/http2_test/test_data_frame_padding.py]: Some comments here will go a long way

File: test/http2_test/http2_base_server.py (hunk 1 above)
Comment (apolcyn) [inline test/http2_test/http2_base_server.py]: done, changed to `_MIN_MAX_FRAME_SIZE`

File: test/http2_test/test_data_frame_padding.py (hunk 2 above)
Comment (apolcyn) [inline test/http2_test/test_data_frame_padding.py]: woops, fixed, this is a mistake.

This tries to resume sending only on the stream which the window update is for, or all if the window update is a connection update - same outcome but mean to limit resuming of sending to only refreshed streams.

File: test/http2_test/test_data_frame_padding.py (hunk 3 above)
Comment (apolcyn) [inline test/http2_test/test_data_frame_padding.py]: added some comments

**Comment by ericgribkoff** [comment]
Could you include a description of the new tests in the http2 test specification document? It's https://github.com/grpc/grpc/blob/master/doc/negative-http2-interop-test-descriptions.md, although it should now be renamed to just `http2-interop-test-descriptions.md` since this PR adds positive tests.

**Comment by adelez** [comment]
Please resolve conflicts first. Interop test show failure to build docker image: https://grpc-testing.appspot.com/job/gRPC_interop_pull_requests/11559/console

File: tools/run_tests/run_interop_tests.py [hunk 4]
```diff
@@ -816,6 +836,11 @@ def aggregate_http2_results(stdout):
                   action='store_const',
                   const=True,
                   help='Enable HTTP/2 server edge case testing. (Good client, bad server)')
+argp.add_argument('--http2_edge_server_interop',
```
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: Can this be combined with http2_badserver_interop and we can call it http2_server_interop? Or do they work with different sets of client languages?

File: tools/run_tests/run_interop_tests.py [hunk 5]
```diff
@@ -982,7 +1011,7 @@ def aggregate_http2_results(stdout):
 
   if args.http2_badserver_interop:
     for language in languages_http2_badserver_interop:
-      for test_case in _HTTP2_BADSERVER_TEST_CASES:
+      for test_case in _HTTP2_BADSERVER_TEST_CASES and test_case not in _POSITIVE_HTTP2_SERVER_TEST_CASES.keys():
```
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: At line 488, positive cases are already added to the negative cases.

File: tools/run_tests/run_interop_tests.py [hunk 6]
```diff
@@ -475,8 +476,17 @@ def __str__(self):
...
+_POSITIVE_HTTP2_SERVER_TEST_CASES = { 'data_frame_padding': 'large_unary', 'no_df_padding_sanity_test': 'large_unary' }
+
+# "badserver_test_cases holds" both "bad server only" tests and the "positive
+# http2 server" tests.
 _HTTP2_BADSERVER_TEST_CASES = ['rst_after_header', 'rst_after_data', 'rst_during_data',
-                     'goaway', 'ping', 'max_streams']
+                               'goaway', 'ping', 'max_streams']
+
+# positive and negative http2 tests are in the same list since they both target the same http2 server.
+_HTTP2_BADSERVER_TEST_CASES.extend(_POSITIVE_HTTP2_SERVER_TEST_CASES.keys())
```
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: In stead of adding positive tests to the negative list, we should create a new var called _HTTP2_SERVER_TEST_CASES.

File: tools/run_tests/run_interop_tests.py [hunk 7]
```diff
@@ -1019,7 +1062,9 @@ def aggregate_http2_results(stdout):
       _LANGUAGES_FOR_HTTP2_BADSERVER_TESTS, resultset, num_failures,
```
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: You need to add these new cases to "http2_badserver_test_cases, _LANGUAGES_FOR_HTTP2_BADSERVER_TESTS" in order for them to show up in the HTML report. 
I suggest renaming these vars by removing "bad" to incorporate both cases. 
You'll also need to modify https://github.com/grpc/grpc/blob/master/tools/run_tests/python_utils/report_utils.py#L83 and https://github.com/grpc/grpc/blob/master/tools/run_tests/interop/interop_html_report.template#L109 accordingly. 
It'll be nice to put them in the same table and call it "HTTP/2 Server Tests" as in https://grpc-testing.appspot.com/job/gRPC_interop_master/HTML_Report/. If neg and pos cases work with different client languages, you can pass a superset of those. I believe if the language is not supported, it'll just say "unimplemented".

File: tools/run_tests/run_interop_tests.py (hunk 4 above)
Comment (apolcyn) [inline tools/run_tests/run_interop_tests.py]: these work with different sets of client languages - but I think they can probably still be combined, WIP on that

**Comment by apolcyn** [comment]
Comments should be addressed now.

@adelez updated the report to show http2 server tests as a superset of positive/negative tests.

Right now, only the core-based languages are failing the "data_frame_padding" tests, but they're passing the "no_df_padding_sanity_test", as expected. go/java are passing both.

I think this is ready for another look now.

**Comment by adelez** [comment]
Looking at the [interop test results](https://grpc-testing.appspot.com/job/gRPC_interop_pull_requests/12193/HTML_Report/), I think we should disable data_frame_padding if some tests fail. An alternative is disabling all languages that fail. File bugs for the failing tests and entered a comment in run_interop_test.py.
Also, non of the negative tests was executed.

File: tools/run_tests/run_interop_tests.py [hunk 8]
```diff
@@ -630,17 +635,29 @@ def cloud_to_cloud_jobspec(language, test_case, server_name, server_host,
...
       '--server_host=%s' % server_host,
   ]
-  if test_case in _HTTP2_BADSERVER_TEST_CASES:
-    # We are running the http2_badserver_interop test. Adjust command line accordingly.
-    offset = sorted(_HTTP2_BADSERVER_TEST_CASES).index(test_case)
+
+  if test_case in _HTTP2_SERVER_TEST_CASES:
+    # We are running the http2_server_interop test. Adjust command line accordingly.
+    offset = sorted(_HTTP2_SERVER_TEST_CASES).index(test_case)
     client_options = common_options + ['--server_port=%s' %
```
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: It seems the negative cases are left out. Maybe that's why they are not executed?

**Comment by apolcyn** [comment]
there were a couple of java to http2 interop failures which look like errors during the tests, but otherwise I think this is ready

File: tools/run_tests/run_interop_tests.py [hunk 9]
```diff
@@ -1049,15 +1071,15 @@ def aggregate_http2_results(stdout):
...
+  if args.http2_server_interop:
     if not args.manual_run:
-      http2_badserver_job.wait_for_healthy(timeout_seconds=600)
-    for language in languages_http2_badserver_interop:
-      for test_case in _HTTP2_BADSERVER_TEST_CASES:
-        offset = sorted(_HTTP2_BADSERVER_TEST_CASES).index(test_case)
+      http2_server_job.wait_for_healthy(timeout_seconds=600)
+    for language in languages_http2_clients_for_http2_server_interop:
+      for test_case in set(_HTTP2_SERVER_TEST_CASES) - set(_HTTP2_SERVER_TEST_CASES_THAT_USE_GRPC_CLIENTS):
+        offset = sorted(_HTTP2_SERVER_TEST_CASES).index(test_case)
```
Comment (ericgribkoff) [inline tools/run_tests/run_interop_tests.py]: This looks broken to me. _HTTP2_SERVER_TEST_CASES now includes `data_frame_padding` and `no_df_padding_sanity_test`. Grabbing the offset from this sorted set will no longer match the 6 test cases implemented by the http2 badserver.

Comment (ericgribkoff) [inline tools/run_tests/run_interop_tests.py]: Ah, nevermind - I missed that these tests are in the http2 (no longer bad) server script as well.

**Comment by ericgribkoff** [comment]
This looks good to me. I can't see any reason for the java http2 interop failures from the changes, but would you mind rerunning the interop tests to be sure? I'm not seeing similar failures on any other PR or on master.

**Comment by apolcyn** [comment]
@ericgribkoff sg re-running

jenkins: test this please

**Comment by adelez** [comment]
The interop result looks good.

**Comment by apolcyn** [comment]
I'll squash these commits down after tests are done

**Comment by apolcyn** [comment]
just squashed and rebased, but also needed to fix sanity test

**Comment by apolcyn** [comment]
test failures:
- mac PR: #9551, #9804 
- windows PR: https://github.com/grpc/grpc/issues/10107
